addons/l10n_account_withholding/wizards/account_payment_register.py

```diff
@@ -0,0 +1,348 @@
+            self.company_id.currency_id,
+            self.company_id,
+            self.payment_date,
+        )
+
+        sign = 1 if self.payment_type == 'inbound' else -1
+
+        # The first step is to add withholding lines.
+        withholding_line_vals = []
+        for withholding_line in self.l10n_account_wth_line_ids:
+            # For each withholding line, we need to create a write_off_line with the values of the tax.
+            withholding_tax_values = withholding_line._get_withholding_tax_values()
+
+            if not withholding_line.name and not withholding_line.tax_id.l10n_account_wth_sequence_id:
+                raise UserError(_('Please enter the withholding number for the tax %(tax_name)s', tax_name=withholding_line.tax_id.name))
+
+            # We set the name of the line for later uses if it's missing but we have a sequence.
+            if not withholding_line.name:
+                withholding_line.name = withholding_line.tax_id.l10n_account_wth_sequence_id.next_by_id()
+
+            # We need to support using multiple tax repartition lines, but in the wizard one tax = one line.
+            # So, we'll only split the result here. If the amount of the line has been set manually, _get_withholding_tax_values already returns the tax details for that amount.
+            for tax in withholding_tax_values:
+                tax_account = tax['account']
+                if not tax_account:
+                    raise UserError(_('Please define a tax account on the distribution of the tax %(tax_name)s', tax_name=tax['tax_name']))
+
+                withholding_line_vals.append({
+                    'currency_id': self.currency_id.id,
+                    'name': f'WH Tax: {withholding_line.name}',
+                    'account_id': tax_account,
+                    'amount_currency': sign * tax['amount'],
+                    'balance': sign * self.company_currency_id.round(tax['amount'] * conversion_rate),
+                    'tax_base_amount': tax['tax_base_amount'],
+                    'tax_repartition_line_id': tax['tax_repartition_line'],
+                    'tax_tag_ids': [Command.set(tax['tag_ids'])],
+                    'analytic_distribution': withholding_line.analytic_distribution,
+                })
+
+        # We also need tax base lines for reporting.
+        # Group the withholding lines by their base amount and base tags.
+        group = self.l10n_account_wth_line_ids.grouped(lambda m: (m.base_amount, self.env['account.account.tag'].browse(m._get_withholding_tax_base_tag_ids())))
+        base_lines_to_create = []
+        # This looping loop aim is to optimize the amount of account.move.line being created by this flow by grouping them.
+        for (group_base_amount, group_tags), withholding_lines in group.items():
+            # This base amount grouped above need to be multiplied in order to correctly affect the tax tags
+            group_base_amount *= len(withholding_lines)
```

Comment:
How does this work with the for loop below?

Reply:
This is quite a vague comment :grimacing: 
Do you have any specific line that you think shouldn't work?

In this case we have grouped the lines per tax tags and amount.
For example:

(100, T1)
(100, T1)
(100, T2)
(100, T2)
(200, T3)
(500, T4)

After grouped:

(100, T1 + T1)
(100, T2 + T2)
(200, T3)
(500, T4)

In the first loop, we need to adapt the base because we "merge" the **tags**

(200, T1)
(200, T2)
(200, T3)
(500, T4)

In the second loop we match on the base, but we don't adapt it again because we don't merge any tags.

(200, T1 + T2 + T3)
(500, T4)